process/process.go

```diff
@@ -0,0 +1,380 @@
+type Scheduler interface {
+	Schedule(block.Height, block.Round) id.Signatory
+}
+
+// A Broadcaster sends a Message to as many Processes in the network as
+// possible.
+type Broadcaster interface {
+	Broadcast(Message)
+}
+
+// A Timer determines the timeout duration at a given Step and `block.Round`.
+type Timer interface {
+	Timeout(step Step, round block.Round) time.Duration
+}
+
+// Processes defines a wrapper type around the []Process type.
+type Processes []Process
+
+// A Process defines a state machine in the distributed replicated state
+// machine. See https://arxiv.org/pdf/1807.04938.pdf for more information.
+type Process struct {
+	mu *sync.Mutex
+
+	signatory  id.Signatory
+	blockchain Blockchain
+	state      State
+
+	proposer    Proposer
+	validator   Validator
+	scheduler   Scheduler
+	broadcaster Broadcaster
+	timer       Timer
+	observer    Observer
+}
+
+// New Process initialised to the default state, starting in the first round.
+func New(signatory id.Signatory, blockchain Blockchain, state State, proposer Proposer, validator Validator, observer Observer, broadcaster Broadcaster, scheduler Scheduler, timer Timer) Process {
+	p := Process{
+		mu: new(sync.Mutex),
+
+		signatory:  signatory,
+		blockchain: blockchain,
+		state:      state,
+
+		proposer:    proposer,
+		validator:   validator,
+		observer:    observer,
+		broadcaster: broadcaster,
+		scheduler:   scheduler,
+		timer:       timer,
+	}
+	p.StartRound(0)
+	return p
+}
+
+// MarshalJSON implements the `json.Marshaler` interface for the Process type,
+// by marshaling its isolated State.
+func (p Process) MarshalJSON() ([]byte, error) {
+	return json.Marshal(p.state)
+}
+
+// UnmarshalJSON implements the `json.Unmarshaler` interface for the Process
+// type, by unmarshaling its isolated State.
+func (p *Process) UnmarshalJSON(data []byte) error {
+	return json.Unmarshal(data, &p.state)
+}
+
+// StartRound is safe for concurrent use. See
+// https://arxiv.org/pdf/1807.04938.pdf for more information.
+func (p *Process) StartRound(round block.Round) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	p.startRound(round)
+}
+
+// HandleMessage is safe for concurrent use. See
+// https://arxiv.org/pdf/1807.04938.pdf for more information.
+func (p *Process) HandleMessage(m Message) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	switch m := m.(type) {
+	case *Propose:
+		p.handlePropose(m)
+	case *Prevote:
+		p.handlePrevote(m)
+	case *Precommit:
+		p.handlePrecommit(m)
+	}
+}
+
+func (p *Process) startRound(round block.Round) {
+	p.state.CurrentRound = round
+	p.state.CurrentStep = StepPropose
+	if p.signatory.Equal(p.scheduler.Schedule(p.state.CurrentHeight, p.state.CurrentRound)) {
+		var proposal block.Block
+		if p.state.ValidBlock.Hash() != block.InvalidHash {
+			proposal = p.state.ValidBlock
+		} else {
+			proposal = p.proposer.BlockProposal(p.state.CurrentHeight, p.state.CurrentRound)
+		}
+		p.broadcaster.Broadcast(NewPropose(
```

Comment:
Just to clarify: We assume that the implementation of the `Broadcaster` will ensure that the current process will get its `Propose` back so that it can broadcast a valid `Prevote` as well. 
If not, it might make sense to call the appropriate `handleXXX()` methods right after broadcasting messages.

Reply:
That is correct, it is currently assumed that the `Broadcaster` implementation will send the `Propose` to its own `Process`. I don't see any risk in making this unnecessary by explicitly calling `handlePropose` (same goes for the others).

Reply:
Implemented.